Global AD indexing and implementation for FV

lindsayad opened the pull request:
This is taken from the middle of #15644 so this may be missing some bug fixes and/or features. Will port as I find that they're missing...

lindsayad committed:
Add configuration option for AD indexing type

lindsayad committed:
Implement global indexing for FV

Refs #15732

lindsayad commented:
If you guys are fine with the configure option, then I will add a recipe (or actually modify the existing Sparse AD opt recipe) that can test the new option, and then add it to this PR.

rwcarlsen commented on framework/configure:
```diff
@@ -1743,6 +1738,55 @@ _ACEOF
+     ;; #(
+  *) :
+    as_fn_error $? "bad value \"$derivative_type\" for --with-derivative-type" "$LINENO" 5 ;;
+esac
```
ahh - my eyes!

lindsayad replied:
Hahaha, luckily the `autoconf` that generates this is super easy to understand :laughing:

rwcarlsen commented on framework/src/fvbcs/FVFluxBC.C:
```diff
@@ -125,6 +125,22 @@ FVFluxBC::computeJacobian(const FaceInfo & fi)
+
+  for (std::size_t i = 0; i < column_indices.size(); ++i)
+    _assembly.cacheJacobianContribution(row_index, column_indices[i], values[i], _matrix_tags);
+#else
```
It seems like this sort of code will be very common and nearly identical for every system/object that ever will use global indexing for jacobians.  Maybe we need to wrap this up in some sort of global helper function or in a helper assembly function or something.  Then every system won't have to have these ifdefs everywhere - they can be consolidated into one place.

rwcarlsen replied:
I guess the challenge is that the global-indexing version will basically be identical for every system/object, but the local/current code for each system is somewhat specialized.  Maybe we have some sort of lambda func that is passed in to the call to the helper function - and the helper function calls the passed in function only if we aren't doing global indexing.

lindsayad replied:
I like that idea. Folks writing `computeJacobian` other than core MOOSE developers would find the lambda scary, but non-core developers shouldn't be overriding `computeJacobian` in general anyways

rwcarlsen commented on framework/src/loops/ComputeJacobianThread.C:
```diff
@@ -336,6 +336,14 @@ ComputeJacobianThread::postElement(const Elem * /*elem*/)
+  // different assembly data members. So we have to call a different addCached method in order to
+  // accumulate the values into the Jacobian
+  if (_fe_problem.haveFV())
+    _fe_problem.addCachedJacobianContributions(_tid);
```
Is there a reason we can't push this ifdef and logic down into FEProblem::addCachedJacobian instead of having an extra function that has to be called everywhere jacobians are being calc'd from?

rwcarlsen commented on framework/src/variables/MooseVariableDataFV.C:
```diff
@@ -802,13 +804,30 @@ MooseVariableDataFV<OutputType>::computeAD(const unsigned int num_dofs, const un
+    // derivative storage container
+    if (_var.kind() == Moose::VAR_NONLINEAR && currently_computing_jacobian)
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+      Moose::derivInsert(_ad_dof_values[i].derivatives(), _dof_indices[i], 1.);
```
Would it be better to include some sort of explicit reset for the derivative values (while retaining the computing jacobian check) - just to make things less dangerous?  After reading this long comment, I'm still not sure I understand how derivatives build up.  Wouldn't the insert operation just see the same global dof index entry and re-insert the 1's to overwrite the existing entries?  Is this problem just because the derivInsert function is "dumb" and doesn't try to match up the inserted indices with existing indices?

lindsayad replied:
There is only one data object per variable. This routine will get called for every element, so by the end of the computeResidual routine, we will have tried to insert every dof index for a given variable. So in the case of CONSTANT MONOMIALS, if I have more than 50 elements, then I'm screwed.

I could explicitly clear the derivative vector when doing residual evaluations when I do the `_ad_dof_values` assignment since that is what is actually done during Jacobian evaluation (behind the scenes in MetaPhysicL)

lindsayad replied:
If for some reason people are doing PJFNK with FV with the default configuration (non-sparse, local indexing) explicit zeroing of the derivative container will be expensive. Perhaps I will just explicitly zero when we are doing global indexing

rwcarlsen reviewed (changes requested):
I'm super excited for this.  What's neat is how relatively straight forward it is - much simpler than the code (I hope) it may/will eventually replace.

joshuahansel commented on framework/src/fvkernels/FVElementalKernel.C:
```diff
@@ -59,17 +59,43 @@ FVElementalKernel::computeResidual()
+  const auto & derivs = r.derivatives();
+
+  const auto & column_indices = derivs.nude_indices();
+  const auto & values = derivs.nude_data();
```
This code is a bit graphic.

joshuahansel commented on framework/src/fvkernels/FVElementalKernel.C:
```diff
@@ -59,17 +59,43 @@ FVElementalKernel::computeResidual()
+  mooseAssert(column_indices.size() == values.size(), "Indices and values size must be the same");
+
+  const auto & row_indices = _var.dofIndices();
+  mooseAssert(row_indices.size() == 1, "We're currently built to use CONSTANT MONOMIALS");
```
Who is "we" in all of these asserts?

rwcarlsen replied:
Yeah - this should read more like "How TF do *you* have more than one dof index with FV only supporting const monomials?".

lindsayad replied:
> Who is "we" in all of these asserts?

you

joshuahansel commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -365,6 +373,17 @@ class MooseVariableFV : public MooseVariableField<OutputType>
+
+private:
+  /**
+   * Check and see whether the AD indexing scheme is compatible with the scaling factors
```
You may have explained this in some discussion, but maybe the basic reasoning should be captured (a comment here would probably suffice).

rwcarlsen replied:
We should at least make a github issue to track progress making this work with scaling.  I expect that this will be VERY important sooner rather than later.  I know with pronghorn at least, scaling is basically required in order to get acceptable convergence.

lindsayad replied:
I've added this to the to-do list for #15732 and beefed up the doxygen string here

joshuahansel commented on framework/src/fvkernels/FVFluxKernel.C:
```diff
@@ -159,7 +175,7 @@ FVFluxKernel::computeJacobian(const FaceInfo & fi)
   _face_info = &fi;
   _normal = fi.normal();
-  const DualReal r = fi.faceArea() * fi.faceCoord() * computeQpResidual();
+  const ADReal r = fi.faceArea() * fi.faceCoord() * computeQpResidual();
```
Are these different, or is this just a style change?

rwcarlsen replied:
They use to be different, but now they are the same ever since Alex slowed AD down :-P

joshuahansel reviewed (changes requested):
I'm guessing you need to add new CIVET recipes before you can add a test for this.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/580474/) on 9052906 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15733/site)

This comment will be updated on new commits.

lindsayad committed:
Move all global indexing logic to Assembly

This allows placement of the ifdef MOOSE_GLOBAL_AD_INDEXING logic to be
moved solely to Assembly; it will not exist in compute objects, like
FVKernels or FVBCs. We take care of the local indexing configuration
case by using lambdas and a LocalFunctor parameter type on
`Assembly::processDerivatives`.

lindsayad commented:
Configure is indeed using the global AD indexing for the Sparse AD opt target. Nice!

lindsayad commented:
@joshuahansel do you want to see if your comments were addressed?

joshuahansel commented:
@lindsayad Were you going to add a test of some sort?

lindsayad commented:
The Sparse AD opt test hits all added code in this PR because we are
replacing the local indexing scheme with our global one.

On Tue, Aug 11, 2020 at 11:16 AM joshuahansel <notifications@github.com>
wrote:

> @lindsayad <https://github.com/lindsayad> Were you going to add a test of
> some sort?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15733#issuecomment-672154722>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4DHLXJTY5KCTEFHPRTSAGDHDANCNFSM4PX2GVOQ>
> .
>

joshuahansel commented:
> The Sparse AD opt test hits all added code in this PR because we are
replacing the local indexing scheme with our global one.

Ok, I see now.

lindsayad commented:
If @rwcarlsen also approves, then I'll squash the fixup commits. (I like having those around as a reviewer so I can see what changed since I last reviewed...)

rwcarlsen commented on framework/include/loops/ComputeFVFluxThread.h:
```diff
@@ -404,7 +404,11 @@ ComputeFVFluxThread<RangeType>::post()
   // make sure we add any remaining cached residuals/jacobians to add/record
   Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
   if (_do_jacobian)
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+    _fe_problem.addCachedJacobianContributions(_tid);
+#else
     _fe_problem.addCachedJacobian(_tid);
+#endif
   else
```
Why can't this be pushed down into FEProblem::addCachedJacobian?

rwcarlsen commented on framework/include/base/Assembly.h:
```diff
@@ -2324,3 +2330,29 @@ Assembly::adGradPhi<RealVectorValue>(const MooseVariableFE<RealVectorValue> & v)
+
+template <typename LocalFunctor>
+void
+Assembly::processDerivatives(const ADReal & residual,
```
This could really use a doc comment explaining when/why/how you use it.

rwcarlsen reviewed (changes requested):
Just a couple of minor things.

moosebuild commented:
All [jobs](https://civet.inl.gov/event/57959/) on 9e16150 : invalidated by @lindsayad

loganharbour commented:
Needs a new commit hash

moosebuild commented:
All [jobs](https://civet.inl.gov/event/57959/) on 9e16150 : invalidated by @lindsayad

rwcarlsen commented:
Why isn't civet status being updated here in a CI panel?

milljm commented:
I needed to hit re-submit on the web hook a couple times...

rwcarlsen commented:
@lindsayad - Are there any PRs that need to go in before this one?

lindsayad commented:
> @lindsayad - Are there any PRs that need to go in before this one?

No sir
